Pzucker/inline sub

philzook58 opened the pull request:
Consolidated mk_default_env and mk_inline_env into a single function.
In addition, any function names starting with "subs_" should be inclined by default

fortunac commented on wp/lib/bap_wp/tests/unit/test_compare.ml:
```diff
@@ -68,8 +68,8 @@ let test_block_pair_1 (test_ctx : test_ctxt) : unit =
   let var_gen = Env.mk_var_gen () in
-  let env1 = Pre.mk_default_env ctx var_gen in
-  let env2 = Pre.mk_default_env ctx var_gen in
+  let env1 = Pre.mk_env ctx var_gen  in
```
Any reason for the extra space at the end?

fortunac commented on wp/plugin/wp.ml:
```diff
@@ -76,18 +81,21 @@ let analyze_proj (proj : project) (var_gen : Env.var_gen) (ctx : Z3.context)
+      let to_inline' = if (inline || has_inline_funcs to_inline) then
+        begin
+          if List.is_empty to_inline then
+            (Printf.printf "using all subs"; subs)
```
Maybe make this in `info` log instead of a print?

fortunac commented on wp/plugin/wp.ml:
```diff
@@ -76,18 +81,21 @@ let analyze_proj (proj : project) (var_gen : Env.var_gen) (ctx : Z3.context)
-      Pre.mk_inline_env ctx var_gen ~subs ~arch ~to_inline:to_inline
-    else
-      Pre.mk_default_env ctx var_gen ~subs ~arch
+      Seq.iter (filter_subs subs) ~f:(fun sub -> Printf.printf "%s\n" (Sub.name sub));
```
What is this supposed to be printing out?

philzook58 commented:
These are good catches

fortunac commented on wp/plugin/wp.ml:
```diff
@@ -134,14 +106,10 @@ let compare_projs (proj : project) (file1: string) (file2 : string)
-    Pre.mk_env ctx var_gen ~subs:subs2 ~arch:arch ~to_inline:(Seq.append to_inline2 (filter_subs subs2))
+    let to_inline1 = match_inline to_inline subs1 in 
+    let to_inline2 = match_inline to_inline subs2 in 
+    Pre.mk_env ctx var_gen ~subs:subs1 ~arch:arch ~to_inline:to_inline1,  (* (Seq.append to_inline1 (filter_subs subs1)) *)
```
Can you remove the commented out code?

philzook58 commented:
Changed to use regular expressions to specify functions to inline
Does add a dependency for the Re library
I changed the inline flag to be the one that takes the regular expression
the old bool taking flag no longer exists
documentation is updated accordingly
and --wp-inline-funcs no longer exists 
--wp-inline=.* will inline everything

fortunac commented on wp/plugin/README.md:
```diff
@@ -9,6 +9,7 @@ Depends on:
     - bap 2.0.0-alpha
     - ounit 2.0.8
     - z3 4.8.6
+    - re 
```
Is there a specific version for `re` we're using?

philzook58 replied:
Added 1.9.0

fortunac commented on wp/plugin/tests/test_wp.ml:
```diff
@@ -107,30 +103,27 @@ let suite = [
-  "Retrowrite Stub: imply inline"  >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~to_inline:"__afl_maybe_log";
-  "Retrowrite Stub: inline all"    >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~inline:true;
+  "Retrowrite Stub"                >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~inline:"__afl_maybe_log";
+  "Retrowrite Stub: imply inline"  >:: test_compare_elf "retrowrite_stub" "UNSAT!" ~inline:"__afl_maybe_log";
```
I think we can remove the "imply inline" tests, now that we consolidated the inline flag.

fortunac commented on wp/plugin/tests/test_wp.ml:
```diff
@@ -107,30 +103,27 @@ let suite = [
-  "Function Spec: inline all "     >:: test_single_elf "function_spec" "main" "UNSAT!" ~inline:true;
-  "Function Spec: no inlining"     >:: test_single_elf "function_spec" "main" "SAT!" ~inline:false;
+  "Function Call"                  >:: test_single_elf "function_call" "main" "SAT!" ~inline:"foo";
+  "Function Call: imply inline"    >:: test_single_elf "function_call" "main" "SAT!" ~inline:"foo";
```
Here too.

fortunac commented on wp/plugin/tests/test_wp.ml:
```diff
@@ -107,30 +103,27 @@ let suite = [
+  "Function Call: imply inline"    >:: test_single_elf "function_call" "main" "SAT!" ~inline:"foo";
+  "Function Call: inline all"      >:: test_single_elf "function_call" "main" "SAT!" ~inline:".*";
+  "Function Spec"                  >:: test_single_elf "function_spec" "main" "UNSAT!" ~inline:"foo";
+  "Function Spec: imply inline"    >:: test_single_elf "function_spec" "main" "UNSAT!" ~inline:"foo";
```
Here too.

philzook58 replied:
Good catch

fortunac commented on wp/plugin/wp.ml:
```diff
@@ -126,14 +106,10 @@ let compare_projs (proj : project) (file1: string) (file2 : string)
-      Pre.mk_default_env ctx var_gen ~subs:subs1 ~arch:arch,
-      Pre.mk_default_env ctx var_gen ~subs:subs2 ~arch:arch
+    let to_inline1 = match_inline to_inline subs1 in 
+    let to_inline2 = match_inline to_inline subs2 in 
```
Do you think it's a good idea to have an info log about which functions are going to be inlined?

philzook58 replied:
Smart.

fortunac commented on wp/plugin/wp.ml:
```diff
@@ -47,48 +47,31 @@ let update_default_num_unroll (num_unroll : int option) : unit =
+let match_inline (to_inline : string option) (subs : (Sub.t Seq.t)) : Sub.t Seq.t =
+  match to_inline with
+  | None -> Seq.empty
+  | Some to_inline -> let inline_pat = Re.Posix.re to_inline |> Re.Posix.compile in
```
Do you think it's a good idea to have a warning log in the case we couldn't match on anything?

philzook58 replied:
Smart.

codyroux replied:
I do!

fortunac commented:
Tests are passing for me, and everything looks super good! :tada:

codyroux commented on wp/lib/bap_wp/tests/unit/test_compare.ml:
```diff
@@ -391,9 +391,9 @@ let test_sub_pair_fun_3 (test_ctx : test_ctxt) : unit =
   let main_sub1 = mk_sub ~tid:sub1_tid ~name:"main_sub" [blk1; blk2; blk3; blk4] in
   let main_sub2 = mk_sub ~tid:sub2_tid ~name:"main_sub" [blk1'; blk2; blk3; blk4] in
-  let env1 = Pre.mk_default_env ctx var_gen
+  let env1 = Pre.mk_env ctx var_gen 
```
Can you install `ocp-indent` pretty please?

codyroux commented:
This looks really great! Good job Phil! Sorry to bug you with more minor comments.
